stream/test.cc

```diff
@@ -840,43 +1021,261 @@ TEST(Stream, ParsesFromFile) {
+
+  const std::string base_url = Printf("http://localhost:%d/exposed", FLAGS_stream_http_test_port);
+  const auto scope = HTTP(FLAGS_stream_http_test_port).Register("/exposed", *exposed_stream);
+
+  const auto CollectSubscriptionResult = [&](std::vector<std::string>& rows, std::vector<std::string>& entries) {
+    rows.clear();
+    entries.clear();
+    RecordsCollector collector(entries, rows);
+    const auto scope = exposed_stream->Subscribe(collector);
+    while (collector.count_ < 3u) {
+      std::this_thread::yield();
+    }
+  };
+  const auto CollectUnsafeSubscriptionResult = [&](std::vector<std::string>& rows, std::vector<std::string>& entries) {
+    rows.clear();
+    entries.clear();
+    RecordsUnsafeCollector collector(entries, rows);
+    const auto scope = exposed_stream->SubscribeUnsafe(collector);
+    while (collector.count_ < 3u) {
+      std::this_thread::yield();
+    }
+  };
+
+  {
+    std::vector<std::string> all_entries;
+    std::vector<std::string> all_rows;
+    const auto expected_rows =
+        "{\"index\":0,\"us\":100}\t{\"x\":1}\n"
+        "{\"index\":1,\"us\":200}\t{\"x\":2}\n"
+        "{\"index\":2,\"us\":400}\t{\"x\":3}\n";
+    const auto expected_entries =
+        "{\"x\":1}\n"
+        "{\"x\":2}\n"
+        "{\"x\":3}\n";
+
+    CollectSubscriptionResult(all_entries, all_rows);
+    EXPECT_EQ(expected_rows, Join(all_rows, ""));
+    EXPECT_EQ(expected_entries, Join(all_entries, ""));
+
+    CollectUnsafeSubscriptionResult(all_entries, all_rows);
+    EXPECT_EQ(expected_rows, Join(all_rows, ""));
+    EXPECT_EQ(expected_entries, Join(all_entries, ""));
+
+    EXPECT_EQ(expected_rows, HTTP(GET(base_url + "?nowait&checked")).body);
+    EXPECT_EQ(expected_rows, HTTP(GET(base_url + "?nowait")).body);
+    EXPECT_EQ(expected_entries, HTTP(GET(base_url + "?nowait&entries_only&checked")).body);
+    EXPECT_EQ(expected_entries, HTTP(GET(base_url + "?nowait&entries_only")).body);
+  }
+
+  {
+    // Swap the first two entries to provoke the `InconsistentIndexException` exceptions.
+    std::swap(lines[0], lines[1]);
+    current::FileSystem::WriteStringToFile(CombineFileContents(), persistence_file_name.c_str());
+
+    std::vector<std::string> entries;
+    std::vector<std::string> rows;
+    const auto expected_rows =
+        "{\"index\":1,\"us\":200}\t{\"x\":2}\n"
+        "{\"index\":0,\"us\":100}\t{\"x\":1}\n"
+        "{\"index\":2,\"us\":400}\t{\"x\":3}\n";
+    const auto expected_entries =
+        "{\"x\":2}\n"
+        "{\"x\":1}\n"
+        "{\"x\":3}\n";
+
+    EXPECT_NO_THROW(CollectUnsafeSubscriptionResult(entries, rows));
+    EXPECT_EQ(expected_rows, Join(rows, ""));
+    EXPECT_EQ(expected_entries, Join(entries, ""));
+    EXPECT_EQ(expected_rows, HTTP(GET(base_url + "?nowait")).body);
+    EXPECT_EQ(expected_entries, HTTP(GET(base_url + "?nowait&entries_only")).body);
+    std::swap(lines[0], lines[1]);
```

Comment:
Looks scary w/o rewriting the file again. Perhaps a comment? Also, if we keep the test this way, why not `swap` the first two lines back right after `WriteStringToFile`, two lines below the first `swap`?